File: nodes/lidar_tf_broadcaster.py
#!/usr/bin/env python
import rospy
import tf
import tf2_ros
from geometry_msgs.msg import TransformStamped
from std_msgs.msg import Float32

AXEL_HEIGHT = 0.02 #  Approximate height of the rotational axis.
BEAM_HEIGHT = 0.10 #  Approximate height of the laser.


def handle_platform_angle(msg):
    #  Initialize the broadcaster.
    br = tf2_ros.TransformBroadcaster()
    
    #  Create and initialize the tf message.
    t = TransformStamped()

    t.header.stamp = rospy.Time.now()
    t.header.frame_id = "lidar_mount"
    t.child_frame_id = "axel"

    #  Convert the angle from the encoder into a quaternion and assign to t.
    quat = tf.transformations.quaternion_from_euler(0, msg.data, 0)
    t.transform.rotation.x = quat[0]
    t.transform.rotation.y = quat[1]
    t.transform.rotation.z = quat[2]
    t.transform.rotation.w = quat[3]
    
    #  Broadcast the transform.
    br.sendTransform(t)


if __name__ == '__main__':
    rospy.init_node('lidar_tf_broadcaster')

    #  Register encoder angle handler.
    rospy.Subscriber('lidar_mount_angle', Float32, handle_platform_angle)
    
    #  Spin forever.
    rospy.spin()

File: firmware/LidarTfBroadcaster.cpp
#include "LidarTfBroadcaster.h"

LidarTfBroadcaster::LidarTfBroadcaster(ros::NodeHandle * node_handle)
{
  nh_ = node_handle;
  br.init(*nh_);
}

void LidarTfBroadcaster::broadcast_transform_from_tilt(float tilt)
{
  geometry_msgs::TransformStamped t;

  t.header.frame_id = "axel";
  t.child_frame_id = "lidar_mount";

  geometry_msgs::Quaternion q;
  q = tf::createQuaternionFromYaw(tilt);

  t.transform.rotation.x = q.x;
  t.transform.rotation.y = q.z; // Swapped because our
  t.transform.rotation.z = q.y; // angle is pitch, not yaw
  t.transform.rotation.w = q.w;
  t.header.stamp = nh_->now();

  br.sendTransform(t);
}
